async_simple/coro/Collect.h

```diff
@@ -123,25 +127,120 @@ struct CollectAnyAwaiter {
             }
 
             input[i].start([i = i, size = input.size(), r = result,
-                            c = continuation,
-                            e = event](Try<ValueType>&& result) mutable {
+                            c = continuation, e = event, callback = _callback](
+                               Try<ValueType>&& result) mutable {
                 assert(e != nullptr);
+                (void)callback;
                 auto count = e->downCount();
                 if (count == size + 1) {
-                    r->_idx = i;
-                    r->_value = std::move(result);
+                    if constexpr (sizeof...(Function) == 0) {
+                        r->_idx = i;
+                        r->_value = std::move(result);
+                    } else {
+                        std::get<0>(callback)(i, std::move(result));
+                    }
                     c.resume();
                 }
             });
         }  // end for
     }
     auto await_resume() {
-        assert(_result != nullptr);
-        return std::move(*_result);
+        if constexpr (sizeof...(Function) == 0) {
+            assert(_result != nullptr);
+            return std::move(*_result);
+        }
     }
 
     std::vector<LazyType, InAlloc> _input;
     std::shared_ptr<ResultType> _result;
+    std::tuple<Function...> _callback;
+};
+
+template <typename... Ts>
+struct CollectAnyVariadicPairAwaiter {
```

Comment:
Can we try to reuse this with `CollectAnyVariadicAwaiter`?

Reply:
The code will become ugly, the input type and return type are different, a new awaiter mabye more clean and clear?

Reply:
Fine.